Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveEPML.cpp

```diff
@@ -0,0 +1,302 @@
+
+    // Loop through the grids, and over the tiles within each grid
+#ifdef _OPENMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+    for ( MFIter mfi(*Efield[0], TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
+
+        // Extract field data for this grid/tile
+        Array4<Real> const& Ex = Efield[0]->array(mfi);
+        Array4<Real> const& Ey = Efield[1]->array(mfi);
+        Array4<Real> const& Ez = Efield[2]->array(mfi);
+#ifndef WARPX_MAG_LLG
+        Array4<Real> const& Bx = Bfield[0]->array(mfi);
+        Array4<Real> const& By = Bfield[1]->array(mfi);
+        Array4<Real> const& Bz = Bfield[2]->array(mfi);
+#endif
+
+        // material macroscopic properties
+        Array4<Real> const& sigma_arr = sigma_mf->array(mfi);
+        Array4<Real> const& eps_arr = eps_mf->array(mfi);
+        Array4<Real> const& mu_arr = mu_mf->array(mfi);
+
+
+        // Extract stencil coefficients
+        Real const * const AMREX_RESTRICT coefs_x = m_stencil_coefs_x.dataPtr();
+        int const n_coefs_x = m_stencil_coefs_x.size();
+        Real const * const AMREX_RESTRICT coefs_y = m_stencil_coefs_y.dataPtr();
+        int const n_coefs_y = m_stencil_coefs_y.size();
+        Real const * const AMREX_RESTRICT coefs_z = m_stencil_coefs_z.dataPtr();
+        int const n_coefs_z = m_stencil_coefs_z.size();
+
+#ifndef WARPX_MAG_LLG
+        FieldAccessorMacroscopic const Hx(Bx, mu_arr);
+        FieldAccessorMacroscopic const Hy(By, mu_arr);
+        FieldAccessorMacroscopic const Hz(Bz, mu_arr);
+#else
+        Array4<Real> const Hx = Hfield[0]->array(mfi);
+        Array4<Real> const Hy = Hfield[1]->array(mfi);
+        Array4<Real> const Hz = Hfield[2]->array(mfi);
+#endif
+
+        // Extract tileboxes for which to loop
+        Box const& tex  = mfi.tilebox(Efield[0]->ixType().toIntVect());
+        Box const& tey  = mfi.tilebox(Efield[1]->ixType().toIntVect());
+        Box const& tez  = mfi.tilebox(Efield[2]->ixType().toIntVect());
+        // starting component to interpolate macro properties to Ex, Ey, Ez locations
+        const int scomp = 0;
+
+        // Loop over the cells and update the fields
+        amrex::ParallelFor(tex, tey, tez,
+
+            [=] AMREX_GPU_DEVICE (int i, int j, int k){
+                // Interpolate conductivity, sigma, to Ex position on the grid
+                amrex::Real const sigma_interp = CoarsenIO::Interp( sigma_arr, sigma_stag,
+                                           Ex_stag, macro_cr, i, j, k, scomp);
+                // Interpolated permittivity, epsilon, to Ex position on the grid
+                amrex::Real const epsilon_interp = CoarsenIO::Interp( eps_arr, epsilon_stag,
+                                           Ex_stag, macro_cr, i, j, k, scomp);
+                amrex::Real alpha = T_MacroAlgo::alpha( sigma_interp, epsilon_interp, dt);
+                amrex::Real beta = T_MacroAlgo::beta( sigma_interp, epsilon_interp, dt);
+
```

Comment:
Comment: the non-macroscopic version doesn't have this alpha/beta part; was sigma assumed to be zero?

Reply:
The alpha part comes from non-zero sigma
The alpha comes from not using c^2, but using the factor function(sigma, epsilon).
This is similar to what we have in MacroscopicEvolveE.cpp

Reply:
Thank you - I understand now.  Forgot that original EvolveE assumed sigma=0.